restore/restore.go

```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		return err
+	}
+	defer gr.Close()
+
+	return r.RestoreService.RestoreShard(ctx, manifest.ShardID, gr)
+}
+
+func (r *restoreRunner) restoreOrganizations(ctx context.Context, req Request) error {
+	var filter influxdb.OrganizationFilter
+	if req.OrgID.Valid() {
+		filter.ID = &req.OrgID
+	}
+	if req.Org != "" {
+		filter.Name = &req.Org
+	}
+
+	orgs, _, err := r.tenantService.FindOrganizations(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	for _, org := range orgs {
+		if err := r.restoreOrganization(ctx, org, req); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *restoreRunner) restoreOrganization(ctx context.Context, org *influxdb.Organization, req Request) error {
+	newOrg := *org
+	if req.NewOrgName != "" {
+		newOrg.Name = req.NewOrgName
+	}
+	r.log.Info(
+		"Restoring organization",
+		zap.String("backup_id", org.ID.String()),
+		zap.String("backup_name", org.Name),
+		zap.String("restored_name", newOrg.Name),
+	)
+
+	// Create organization on server if it doesn't already exist.
+	if o, err := r.OrgService.FindOrganization(ctx, influxdb.OrganizationFilter{Name: &newOrg.Name}); influxdb.ErrorCode(err) == influxdb.ENotFound {
+		if err := r.OrgService.CreateOrganization(ctx, &newOrg); err != nil {
+			return fmt.Errorf("failed to create organization %q: %w", newOrg.Name, err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to check existence of organization %q: %w", newOrg.Name, err)
+	} else {
+		newOrg.ID = o.ID
+	}
+
+	filter := influxdb.BucketFilter{OrganizationID: &newOrg.ID}
+	if req.BucketID.Valid() {
+		filter.ID = &req.BucketID
+	}
+	if req.Bucket != "" {
+		filter.Name = &req.Bucket
+	}
+
+	buckets, _, err := r.tenantService.FindBuckets(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	for _, bkt := range buckets {
+		// Skip internal buckets.
+		if strings.HasPrefix(bkt.Name, "_") {
+			continue
+		}
+
+		bkt = bkt.Clone()
+		bkt.OrgID = newOrg.ID
+
+		if err := r.restoreBucket(ctx, bkt, req); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *restoreRunner) restoreBucket(ctx context.Context, bkt *influxdb.Bucket, req Request) error {
+	newBucket := *bkt
+	if req.NewBucketName != "" {
+		newBucket.Name = req.NewBucketName
+	}
+	r.log.Info(
+		"Restoring bucket",
+		zap.String("backup_id", bkt.ID.String()),
+		zap.String("backup_name", bkt.Name),
+		zap.String("restored_name", newBucket.Name),
+	)
+
+	// Lookup matching database from the meta store.
+	// Search using bucket ID from backup.
+	dbi := r.metaClient.Database(bkt.ID.String())
+	if dbi == nil {
+		return fmt.Errorf("database for bucket %q not found in local backup", bkt.ID.String())
+	}
+
+	// Serialize to protobufs.
+	buf, err := dbi.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("failed to marshal database info for bucket %q: %w", bkt.ID.String(), err)
+	}
+
+	// TODO: Only create if it doesn't already exist.
+	if err := r.BucketService.CreateBucket(ctx, &newBucket); err != nil {
```

Comment:
Assuming that CreateBucket() doesn't fail if the bucket exists

Reply:
CreateBucket() does fail in that case (see #20758). I wanted to be bug-for-bug compatible for the refactor, and fix the issues in follow-up PRs